src/channel_internal.c

```diff
@@ -0,0 +1,395 @@
+        /*Codes_SRS_CHANNEL_INTERNAL_43_123: [ channel_internal_push shall set *out_op_push to the created THANDLE(ASYNC_OP). ]*/
+        THANDLE_INITIALIZE_MOVE(ASYNC_OP)(out_op, &async_op);
+        result = 0;
+    }
+    return result;
+}
+
+static int dequeue_operation(CHANNEL_INTERNAL* channel_internal, THANDLE(ASYNC_OP)* out_op, PULL_CALLBACK pull_callback, void* pull_context, PUSH_CALLBACK push_callback, void* push_context, THANDLE(RC_PTR) data)
+{
+    int result;
+
+    /*Codes_SRS_CHANNEL_INTERNAL_43_109: [ channel_internal_pull shall call DList_RemoveHeadList on the list of pending operations to obtain the operation. ]*/
+    /*Codes_SRS_CHANNEL_INTERNAL_43_125: [ channel_internal_push shall call DList_RemoveHeadList on the list of pending operations to obtain the operation. ]*/
+    PDLIST_ENTRY op_entry = DList_RemoveHeadList(&channel_internal->op_list);
+
+    CHANNEL_OP* channel_op = CONTAINING_RECORD(op_entry, CHANNEL_OP, anchor);
+    if (pull_callback)
+    {
+        /*Codes_SRS_CHANNEL_INTERNAL_43_112: [ channel_internal_pull shall store the pull_callback and pull_context in the obtained operation. ]*/
+        channel_op->pull_callback = pull_callback;
+        channel_op->pull_context = pull_context;
+    }
+    else if (push_callback)
+    {
+        /*Codes_SRS_CHANNEL_INTERNAL_43_128: [ channel_internal_push shall store the push_callback, push_context and data in the obtained operation. ]*/
+        channel_op->push_callback = push_callback;
+        channel_op->push_context = push_context;
+        THANDLE_ASSIGN(RC_PTR)(&channel_op->data, data);
+    }
+
+    /*Codes_SRS_CHANNEL_INTERNAL_43_114: [ channel_internal_pull shall set *out_op_pull to the THANDLE(ASYNC_OP) of the obtained operation. ] ]*/
+    /*Codes_SRS_CHANNEL_INTERNAL_43_130: [ channel_internal_push shall set *out_op_push to the THANDLE(ASYNC_OP) of the obtained operation. ]*/
+    THANDLE_INITIALIZE(ASYNC_OP)(out_op, channel_op->async_op);
+
+    /*Codes_SRS_CHANNEL_INTERNAL_43_113: [ channel_internal_pull shall call threadpool_schedule_work with execute_callbacks as work_function and the obtained operation as work_function_context. ]*/
+    /*Codes_SRS_CHANNEL_INTERNAL_43_129: [ channel_internal_push shall call threadpool_schedule_work with execute_callbacks as work_function and the obtained operation as work_function_context. ]*/
+    if (threadpool_schedule_work(channel_internal->threadpool, execute_callbacks, channel_op) != 0)
+    {
+        LogError("Failure in threadpool_schedule_work(execute_callbacks, channel_op)");
+        THANDLE_ASSIGN(ASYNC_OP)(out_op, NULL);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        result = 0;
+    }
+    return result;
+}
+
+
+IMPLEMENT_MOCKABLE_FUNCTION(, CHANNEL_RESULT, channel_internal_pull, THANDLE(CHANNEL_INTERNAL), channel_internal, PULL_CALLBACK, pull_callback, void*, pull_context, THANDLE(ASYNC_OP)*, out_op_pull)
```

Comment:
>CHANNEL_RESULT, channel_internal_pull [](http://example.com/codeflow?start=30&length=37)

This function doesn't check it's parameters, bad things can happen.

Reply:
arg validation is done by `channel`.